File: scripts/PyMWIR/ForwardRenderer.py
import torch
from . import PyBindMWIR
from . import Scene

class ForwardRenderer:
    def __init__(self):
        self.forward_renderer = PyBindMWIR.ForwardRenderer()

    def Render(self, scene, output=None):
        if self.forward_renderer is None:
            raise ValueError("InverseRenderer ownership has been transferred")
        if type(scene) is not Scene:
            raise TypeError("scene must be of type Scene")
        if output is not None and type(output) is not torch.Tensor:
            raise TypeError("output must be a torch.Tensor or None")
        return self.forward_renderer.Render(scene.scene, output)

File: src/pybindmwir/forward_renderer.cpp
#include "pybindmwir/forward_renderer.hpp"

namespace py = pybind11;


ForwardRenderer::ForwardRenderer()
{
    mwir_renderer_ = std::make_unique<MWIR::ForwardRenderer>();
}
 
torch::Tensor ForwardRenderer::Render(std::shared_ptr<Scene> scene, std::optional<torch::Tensor> result_tensor)
{
    if (!mwir_renderer_)
    {
        throw std::runtime_error("ForwardRenderer ownership has been transferred.");
    }

    return mwir_renderer_->Render(*scene->mwir_scene_, result_tensor);
}


void init_forward_renderer(py::module_ &m)
{
    
    py::class_<ForwardRenderer, std::shared_ptr<ForwardRenderer>>(m, "ForwardRenderer")
        .def(py::init<>())
        .def("Render", &ForwardRenderer::Render);

}
